sui_programmability/examples/defi/sources/pool.move

```diff
@@ -0,0 +1,309 @@
+// Copyright (c) 2022, Mysten Labs, Inc.
+// SPDX-License-Identifier: Apache-2.0
+
+/// Implementation of a liquidity Pool for Sui.
+///
+/// This solution is rather simple and is based on the example from the Move repo:
+/// https://github.com/move-language/move/blob/main/language/documentation/examples/experimental/coin-swap/sources/CoinSwap.move
+module defi::pool {
+    use sui::Coin::{Self, Coin, TreasuryCap};
+    use sui::Balance::{Self, Balance};
+    use sui::ID::{VersionedID};
+    use sui::SUI::SUI;
+    use sui::Transfer;
+    use sui::TxContext::{Self, TxContext};
+
+    /// The Capability to create new pools.
+    struct PoolCreatorCap has key, store {
+        id: VersionedID
+    }
+
+    /// The Pool token that will be used to mark the pool share
+    /// of a liquidity provider.
+    struct LSP<phantom T> has drop {}
+
+    /// The pool with exchange.
+    struct Pool<phantom T> has key {
+        id: VersionedID,
+        sui: Balance<SUI>,
+        token: Balance<T>,
+        lsp_treasury: TreasuryCap<LSP<T>>
+    }
+
+    /// On init creator of the module gets the capability
+    /// to create new `Pool`s.
+    fun init(ctx: &mut TxContext) {
+        Transfer::transfer(PoolCreatorCap {
+            id: TxContext::new_id(ctx)
+        }, TxContext::sender(ctx))
+    }
+
+    /// Create new `Pool` for token `T`. Each Pool holds a `Coin<T>`
+    /// and a `Coin<SUI>`. Swaps are available in both directions.
+    ///
+    /// - `share` argument defines the initial amount of LSP tokens
+    /// received by the creator of the Pool.
+    public fun create_pool<T>(
+        _: &PoolCreatorCap,
+        token: Coin<T>,
+        sui: Coin<SUI>,
+        share: u64,
```

Comment:
Kind of confused by this part. What would go wrong if this created an empty pool and created no LSP tokens, and the only way to get LSP tokens was via `add_liquidity`?

Reply:
We need this initializer for setting LSP proportion. I guess the best solution would be adding an assertion for zeros.

Reply:
What about explicitly storing the proportion in the `Pool`? I think that would let you avoid creating LSP tokens here + the creator will also not need to supply `token` and `sui`. And then there is only one way to add liquidity to the pool (`add_liquidity`).

Reply:
@sblackshear 

I don't think I get it. People need to get something to mark their share of tokens. And there has to be an initial proportion set. Otherwise they will have nothing to offer in `remove_liquidity`.

Reply:
Ok, sorry, I think I get it now. For concreteness, here is what I was concerned about:

Scenario 1:
- Someone calls `create_pool(ten_tokens, ten_sui, 10, 0)`  and gets back 10 LSP tokens
- Someone else calls `add_liquidity(ten_tokens, ten_sui)` and gets back 10 * 10 / 10 = 10 LSP tokens

Scenario 2
- Someone calls `create_pool(ten_tokens, ten_sui, 20, 0)`  and gets back 20 LSP tokens
- Someone else calls `add_liquidity(ten_tokens, ten_sui)` and gets back 10 * 20 / 10 = 20 LSP tokens

I was thinking that in both cases, the caller of `add_liquidity` should get the same *absolute* number of LSP tokens. But that's not what's important here--the key property is that the caller of `add_liquidity` should get the same number of LSP tokens in the caller of `create_pool`. It's just that in the second case, each individual LSP token is worth less.

Reply:
Are you suggesting we set the value of LSP tokens based on SUI amount? I just realized that we don't want LSP value set too low, otherwise due to rounding error someone might get 19 instead of 20 if they sent a little bit smaller value.